src/main/java/bio/terra/cli/service/utils/WorkspaceManagerService.java

```diff
@@ -214,14 +220,52 @@ public void deleteWorkspace(UUID workspaceId) {
    */
   public void grantIamRole(UUID workspaceId, String userEmail, IamRole iamRole) {
     WorkspaceApi workspaceApi = new WorkspaceApi(apiClient);
+    GrantRoleRequestBody grantRoleRequestBody = new GrantRoleRequestBody().memberEmail(userEmail);
     try {
-      GrantRoleRequestBody grantRoleRequestBody = new GrantRoleRequestBody().memberEmail(userEmail);
       workspaceApi.grantRole(grantRoleRequestBody, workspaceId, iamRole);
-    } catch (Exception ex) {
-      logger.error("Error granting IAM role on workspace", ex);
+    } catch (ApiException ex) {
+      // a bad request is the only type of exception that inviting the user might fix
+      if (!isBadRequest(ex)) {
+        throw new RuntimeException("Error granting IAM role on workspace", ex);
+      }
+
+      try {
+        // try to invite the user first, in case they are not already registered
+        // if they are already registered, this will throw an exception
+        logger.info("inviting new user: {}", userEmail);
+        UserStatusDetails userStatusDetails =
+            new SamService(server, terraUser).inviteUser(userEmail);
+        logger.info("invited new user: {}", userStatusDetails);
+
+        // now try to add the user to the workspace role
+        // retry if it returns with a bad request (because the invite sometimes takes a few seconds
+        // to propagate -- not sure why)
+        HttpUtils.callWithRetries(
+            () -> {
+              workspaceApi.grantRole(grantRoleRequestBody, workspaceId, iamRole);
```

Comment:
just curious, I think what's here is fine. Sam calls are always wrapped in retries, but not WM API calls, except for this case. Is Sam just unreliable and WM reliable?

Reply:
- I wrapped this case in retries because when I first tested this code, it fails on the first grantRole call after inviting a user, but succeeds a few seconds later. So it was to spot fix this particular case.
- I think any time we're talking to a service, we should use retries, including for WM. I just haven't added it yet because, as you say, it's been much more reliable than SAM, so it's not been a problem. I added a [ticket](https://broadworkbench.atlassian.net/browse/PF-513) so we don't forget about this. Also probably good to check with WM team what errors should be retryable.